extensions/refresh-catalog/src/main/java/org/eclipse/dataspaceconnector/mvd/FederatedCacheNodeResolver.java

```diff
@@ -0,0 +1,66 @@
+/*
+ *  Copyright (c) 2022 Microsoft Corporation
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+
+package org.eclipse.dataspaceconnector.mvd;
+
+import org.eclipse.dataspaceconnector.catalog.spi.FederatedCacheNode;
+import org.eclipse.dataspaceconnector.iam.did.spi.document.DidDocument;
+import org.eclipse.dataspaceconnector.iam.did.spi.document.Service;
+import org.eclipse.dataspaceconnector.iam.did.spi.resolution.DidResolverRegistry;
+import org.eclipse.dataspaceconnector.registration.client.models.Participant;
+import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.Result;
+
+import java.util.List;
+import java.util.Optional;
+
+import static java.lang.String.format;
+
+/**
+ * Resolves the {@link FederatedCacheNode}s from the Participant's Did Document.
+ */
+class FederatedCacheNodeResolver {
+
+    public static final String IDS_MESSAGING = "IDSMessaging";
+    public static final List<String> IDS_MULTIPART = List.of("ids-multipart");
+
+    private final DidResolverRegistry resolver;
+    private final Monitor monitor;
+
+    FederatedCacheNodeResolver(DidResolverRegistry resolver, Monitor monitor) {
+        this.resolver = resolver;
+        this.monitor = monitor;
+    }
+
+    public Result<FederatedCacheNode> toFederatedCacheNode(Participant participant) {
+        String did = participant.getDid();
+        monitor.debug(format("Resolving Did Document for did %s.", did));
+        Result<DidDocument> didDocument = resolver.resolve(did);
+        if (didDocument.failed()) {
+            monitor.severe(() -> format("Failed to resolve DID Document for %s. %s", did, didDocument.getFailureDetail()));
+            return Result.failure("Can't resolve Did Document for participant: " + did);
+        }
+        return getUrl(didDocument)
+                .map(url -> Result.success(new FederatedCacheNode(didDocument.getContent().getId(), url, IDS_MULTIPART)))
+                .orElseGet(() -> Result.failure(format("Can't resolve Did Document for participant: %s", did)));
+    }
+
+    private Optional<String> getUrl(Result<DidDocument> didDocument) {
```

Comment:
Methods should have neither `Optional` nor `Result` in signatures. I'm assuming this was introduced by the IDE's extract-method feature

Reply:
@paullatzelsperger I understand for `Optional`, but what do you mean about not returning `Result`? Isn't that what Result is for?

Reply:
As I understand it's about not having a Result as a method argument, not about the return type?

Reply:
Changed to not use Result as a method argument

Reply:
yeah @algattik @Izzzu, sorry I formulated that one unclear :)
of course methods can have `Result` as return type.